source/language/scope.rst

```diff
@@ -0,0 +1,318 @@
+behave for scoping purposes as if they were defined in the scope of the
+definition.  The lifetime of these local variables ends at the end of the
+function body, and they are not accessible after the subroutine or gate body.
+Similarly, the qubit identifiers in a gate definition are valid only within the
+definition of the gate.
+
+The identifier of a subroutine or gate is available in the scope of its own
+body, allowing direct recursion.  For gates, the direct recursion is unlikely to
+ever be useful, since this would generally be non-terminating.
+
+Local subroutine or gate variables, including parameters and qubit definitions,
+may shadow variables defined in the outer scope.  Inside the body, the
+identifier will refer to the local variable instead.  After the definition of
+the body has completed (and we are back in the global scope), the identifier
+will refer to the same variable it did before the subroutine or gate.
+
+Subroutines cannot contain ``qubit`` declarations in their bodies, but can
+accept variables of type ``qubit`` in their parameter lists.  Aliases can be
+declared within subroutine and gate scopes, and have the same lifetime and
+visibility as other local variables.
+
+For example:
+
+.. code-block:: c
+   :linenos:
+
+   OPENQASM 3.0;
+
+   qubit[5] all_qubits;
+
+   int a = 1;
+   int b = 2;
+   const int c = 3;
+   const int d = 4;
+
+   def my_routine(uint a, uint c) {
+      // In this body, 'a' refers to the subroutine parameter, not the external
+      // variable, which wouldn't be visible even without the shadowing.
+
+      int local = 5;
+
+      // Identifiers in scope are:
+      //  - 'my_routine': the subroutine itself
+      //  - 'a': type 'uint', from the parameter list
+      //  - 'c': type 'uint', from the parameter list (shadows the outer 'const
+      //      int' 'c').
+      //  - 'd': type 'const int', value 4, visible from the global scope
+      //      because it is a 'const' type.
+      //  - 'local': type 'int', value 5, from regular definition in the current
+      //      scope.
+      //  - other built-in identifiers (such as 'U' and 'pi') that are
+      //      implicitly defined in the global scope.
+      //  - all available hardware qubits (such as '$0')
+      //
+      // The variable 'b' is not in scope, because its visibility as a
+      // non-'const' type does not make it available within subroutines.  The
+      // hardware qubit identifiers are in scope, but not the virtual qubit
+      // identifier 'all_qubits'.
+   }
+
+   // After the subroutine block, 'a' and 'c' once again refer to the variables
+   // of type 'int' and 'const int' defined on lines 3 and 5 respectively.
+   // 'local' (from the subroutine body) is not in scope, while 'my_routine'
+   // (the subroutine) is.
+
+   const float[64] new_variable = 1.5;
+
+   def second_subroutine(qubit[4] q) {
+      int local = 8;
+
+      let alias = q[0:2];
```

Comment:
nitpick: Use a different name than alias to disambiguate the variable and the meaning.

Reply:
Done in dbc451c.